File: go/internal/domains/ai_workflow/ai_workflow.go
// Package ai_workflow — multi-step pipelines over a running value: each step's output threads into the next.
// The dangerous property is TERMINATION + CONTAINMENT: a run ALWAYS terminates (MAX_STEPS bounds every run) and
// a failing or unknown step is CONTAINED (the run stops gracefully with ok:false and the trace so far — never a
// crash, never a 5xx). String ops slice and measure by CODEPOINTS (runes) — the ×3-identical semantic.
// Definitions are durable. Store names and shapes match the python/node impls.
package ai_workflow

import (
	"net/http"
	"strconv"

	"app/internal/core"
)

const aiWorkflowMaxSteps = 50

type aiWorkflowStep = map[string]any

type aiWorkflowDef struct {
	Id    int              `json:"id"`
	Steps []aiWorkflowStep `json:"steps"`
}

var aiWorkflowDefs = core.NewKV[string, aiWorkflowDef]("ai_workflow_defs")

// one step. Returns (newValue, ok) — unknown/invalid ops report ok=false, they never panic.
func aiWorkflowApply(op, value string, step aiWorkflowStep) (string, bool) {
	text, _ := step["text"].(string)
	switch op {
	case "append":
		return value + text, true
	case "prepend":
		return text + value, true
	case "truncate":
		n := 0
		if f, isNum := step["n"].(float64); isNum && f >= 0 {
			n = int(f)
		}
		runes := []rune(value) // CODEPOINT slicing — parity with python/node
		if n > len(runes) {
			n = len(runes)
		}
		return string(runes[:n]), true
	case "length":
		return strconv.Itoa(len([]rune(value))), true // CODEPOINT count
	}
	return value, false
}

func AiWorkflowCreate(w http.ResponseWriter, r *http.Request) {
	// decode FIRST: DecodeJSON enforces the body cap (413) and drains the stream — replying (incl. a 401) before
	// the body is read aborts the connection mid-upload. Identity is checked next, before any write. (×3 parity)
	in, ok := core.DecodeJSON[struct {
		Steps []aiWorkflowStep `json:"steps"`
	}](w, r)
	if !ok {
		return
	}
	if _, ok := core.RequireIdentity(w, r); !ok { // authenticated mutation (no/invalid token -> 401)
		return
	}
	if len(in.Steps) == 0 {
		core.WriteProblem(w, 422, "a workflow needs at least one step")
		return
	}
	for _, step := range in.Steps {
		op, isStr := step["op"].(string)
		if step == nil || !isStr || op == "" {
			core.WriteProblem(w, 422, "every step must be an object with a string 'op'")
			return
		}
	}
	wid := core.NextID("ai_workflow_def")
	aiWorkflowDefs.Set(strconv.Itoa(wid), aiWorkflowDef{Id: wid, Steps: in.Steps})
	core.WriteJSON(w, 201, map[string]any{"id": wid, "steps": len(in.Steps)})
}

func AiWorkflowRun(w http.ResponseWriter, r *http.Request) {
	if _, ok := core.RequireIdentity(w, r); !ok { // identity before the path id (×3: a no-token float-path -> 401)
		return
	}
	wid, err := strconv.Atoi(r.PathValue("workflow_id"))
	if err != nil {
		core.WriteProblem(w, 422, "invalid workflow id")
		return
	}
	in, ok := core.DecodeJSON[struct {
		Input *string `json:"input"`
	}](w, r)
	if !ok {
		return
	}
	if in.Input == nil {
		core.WriteProblem(w, 422, "invalid body")
		return
	}
	wf, exists := aiWorkflowDefs.Get(strconv.Itoa(wid))
	if !exists {
		core.WriteProblem(w, 404, "workflow not found")
		return
	}
	value, runOk := *in.Input, true
	trace := []map[string]string{}
	limit := len(wf.Steps)
	if limit > aiWorkflowMaxSteps {
		limit = aiWorkflowMaxSteps // TERMINATION: never more than MAX_STEPS, whatever was defined
	}
	for _, step := range wf.Steps[:limit] {
		op := step["op"].(string)
		next, stepOk := aiWorkflowApply(op, value, step)
		value = next
		if !stepOk {
			runOk = false // CONTAINMENT: stop gracefully, keep the trace so far
			break
		}
		trace = append(trace, map[string]string{"op": op, "output": value})
	}
	if runOk && len(wf.Steps) > aiWorkflowMaxSteps {
		runOk = false // the budget itself was exceeded — report it, loudly
	}
	core.WriteJSON(w, 200, map[string]any{"output": value, "steps_run": len(trace), "ok": runOk, "trace": trace})
}

File: node/src/domains/ai_workflow.js
// ai_workflow — multi-step pipelines over a running value: each step's output threads into the next. The
// dangerous property is TERMINATION + CONTAINMENT: a run ALWAYS terminates (MAX_STEPS bounds every run) and a
// failing or unknown step is CONTAINED (the run stops gracefully with ok:false and the trace so far — never a
// crash, never a 5xx). String ops slice and measure by CODEPOINTS (spread, not .length) — the x3-identical
// semantic. Definitions are durable. Store names and shapes match the python/go impls.
import { intParam, nextId, problem, requireIdentity, sendJSON, storeGet, storePut } from '../core/runtime.js';

const MAX_STEPS = 50;
// state in store: seq "ai_workflow_def" · ns "ai_workflow_defs" String(id) -> {id, steps}

// one step. Returns [newValue, ok] — unknown/invalid ops report ok=false, they never throw.
function apply(op, value, step) {
  const text = typeof step.text === 'string' ? step.text : '';
  if (op === 'append') return [value + text, true];
  if (op === 'prepend') return [text + value, true];
  if (op === 'truncate') {
    const n = Number.isInteger(step.n) && step.n >= 0 ? step.n : 0;
    return [[...value].slice(0, n).join(''), true]; // CODEPOINT slicing — parity with python/go
  }
  if (op === 'length') return [String([...value].length), true]; // CODEPOINT count
  return [value, false];
}

export async function aiWorkflowCreate(req, res, params, body) {
  if ((await requireIdentity(req, res)) === null) return; // authenticated mutation (no/invalid token -> 401)
  if (!body || !Array.isArray(body.steps) || body.steps.length === 0) {
    return problem(res, 422, 'a workflow needs at least one step');
  }
  for (const step of body.steps) {
    if (!step || typeof step !== 'object' || Array.isArray(step) || typeof step.op !== 'string' || step.op === '') {
      return problem(res, 422, "every step must be an object with a string 'op'");
    }
  }
  const wid = await nextId('ai_workflow_def');
  await storePut('ai_workflow_defs', String(wid), { id: wid, steps: body.steps });
  sendJSON(res, 201, { id: wid, steps: body.steps.length });
}

export async function aiWorkflowRun(req, res, params, body) {
  if ((await requireIdentity(req, res)) === null) return; // authenticated mutation (no/invalid token -> 401)
  const wid = intParam(params.workflow_id);
  if (wid === null) return problem(res, 422, 'invalid workflow id');
  if (!body || typeof body.input !== 'string') return problem(res, 422, 'invalid body');
  const wf = await storeGet('ai_workflow_defs', String(wid));
  if (wf === undefined) return problem(res, 404, 'workflow not found');
  let value = body.input;
  let ok = true;
  const trace = [];
  // TERMINATION: never more than MAX_STEPS, whatever was defined
  for (const step of wf.steps.slice(0, MAX_STEPS)) {
    const [next, stepOk] = apply(step.op, value, step);
    value = next;
    if (!stepOk) { ok = false; break; } // CONTAINMENT: stop gracefully, keep the trace so far
    trace.push({ op: step.op, output: value });
  }
  if (ok && wf.steps.length > MAX_STEPS) ok = false; // the budget itself was exceeded — report it, loudly
  sendJSON(res, 200, { output: value, steps_run: trace.length, ok, trace });
}
